## **Amazon Sales Data Analysis: Insights into Product Ratings, Pricing, and Customer Reviews**

### **Project Objective:**
To analyze the Amazon Sales Dataset to uncover insights about product ratings, pricing, discounts, and customer reviews across different categories. The analysis will help identify trends, popular products, and factors influencing customer satisfaction.

### **Import Libraries**

In [164]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

In [165]:
# Load the Data from the CSV file

amazon_sales = pd.read_csv("Data Source/amazon.csv")
amazon_sales.head(5)

product_id                                       product_name  \
0  B07JW9H4J1  Wayona Nylon Braided USB to Lightning Fast Cha...   
1  B098NS6PVG  Ambrane Unbreakable 60W / 3A Fast Charging 1.5...   
2  B096MSW6CT  Sounce Fast Phone Charging Cable & Data Sync U...   
3  B08HDJ86NZ  boAt Deuce USB 300 2 in 1 Type-C & Micro USB S...   
4  B08CF3B7N1  Portronics Konnect L 1.2M Fast Charging 3A 8 P...   

                                            category discounted_price  \
0  Computers&Accessories|Accessories&Peripherals|...             ₹399   
1  Computers&Accessories|Accessories&Peripherals|...             ₹199   
2  Computers&Accessories|Accessories&Peripherals|...             ₹199   
3  Computers&Accessories|Accessories&Peripherals|...             ₹329   
4  Computers&Accessories|Accessories&Peripherals|...             ₹154   

  actual_price discount_percentage rating rating_count  \
0       ₹1,099                 64%    4.2       24,269   
1         ₹349                 43%    4.0       43,994   
2       ₹1,899                 90%    3.9        7,928   
3         ₹699                 53%    4.2       94,363   
4         ₹399                 61%    4.2       16,905   

                                       about_product  \
0  High Compatibility : Compatible With iPhone 12...   
1  Compatible with all Type C enabled devices, be...   
2  【 Fast Charger& Data Sync】-With built-in safet...   
3  The boAt Deuce USB 300 2 in 1 cable is compati...   
4  [CHARGE & SYNC FUNCTION]- This cable comes wit...   

                                             user_id  \
0  AG3D6O4STAQKAY2UVGEUV46KN35Q,AHMY5CWJMMK5BJRBB...   
1  AECPFYFQVRUWC3KGNLJIOREFP5LQ,AGYYVPDD7YG7FYNBX...   
2  AGU3BBQ2V2DDAMOAKGFAWDDQ6QHA,AESFLDV2PT363T2AQ...   
3  AEWAZDZZJLQUYVOVGBEUKSLXHQ5A,AG5HTSFRRE6NL3M5S...   
4  AE3Q6KSUK5P75D5HFYHCRAOLODSA,AFUGIFH5ZAFXRDSZH...   

                                           user_name  \
0  Manav,Adarsh gupta,Sundeep,S.Sayeed Ahmed,jasp...   
1  ArdKn,Nirbhay kumar,Sagar Viswanathan,Asp,Plac...   
2  Kunal,Himanshu,viswanath,sai niharka,saqib mal...   
3  Omkar dhale,JD,HEMALATHA,Ajwadh a.,amar singh ...   
4  rahuls6099,Swasat Borah,Ajay Wadke,Pranali,RVK...   

                                           review_id  \
0  R3HXWT0LRP0NMF,R2AJM3LFTLZHFO,R6AQJGUP6P86,R1K...   
1  RGIQEG07R9HS2,R1SMWZQ86XIN8U,R2J3Y1WL29GWDE,RY...   
2  R3J3EQQ9TZI5ZJ,R3E7WBGK7ID0KV,RWU79XKQ6I1QF,R2...   
3  R3EEUZKKK9J36I,R3HJVYCLYOY554,REDECAZ7AMPQC,R1...   
4  R1BP4L2HH9TFUP,R16PVJEXKV6QZS,R2UPDB81N66T4P,R...   

                                        review_title  \
0  Satisfied,Charging is really fast,Value for mo...   
1  A Good Braided Cable for Your Type C Device,Go...   
2  Good speed for earlier versions,Good Product,W...   
3  Good product,Good one,Nice,Really nice product...   
4  As good as original,Decent,Good one for second...   

                                      review_content  \
0  Looks durable Charging is fine tooNo complains...   
1  I ordered this cable to connect my phone to An...   
2  Not quite durable and sturdy,https://m.media-a...   
3  Good product,long wire,Charges good,Nice,I bou...   
4  Bought this instead of original apple, does th...   

                                            img_link  \
0  https://m.media-amazon.com/images/W/WEBP_40237...   
1  https://m.media-amazon.com/images/W/WEBP_40237...   
2  https://m.media-amazon.com/images/W/WEBP_40237...   
3  https://m.media-amazon.com/images/I/41V5FtEWPk...   
4  https://m.media-amazon.com/images/W/WEBP_40237...   

                                        product_link  
0  https://www.amazon.in/Wayona-Braided-WN3LG1-Sy...  
1  https://www.amazon.in/Ambrane-Unbreakable-Char...  
2  https://www.amazon.in/Sounce-iPhone-Charging-C...  
3  https://www.amazon.in/Deuce-300-Resistant-Tang...  
4  https://www.amazon.in/Portronics-Konnect-POR-1...

### **Data Exploration**

In [166]:
# Step 1: Data Exploration

# no of rows and columns in dataset
print('\nShape of the Dataset: ')
print(amazon_sales.shape)

# Overview of Dataset
print("\nOverview of Dataset: ")
print(amazon_sales.info())



Shape of the Dataset: 
(1465, 16)

Overview of Dataset: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   product_id           1465 non-null   object
 1   product_name         1465 non-null   object
 2   category             1465 non-null   object
 3   discounted_price     1465 non-null   object
 4   actual_price         1465 non-null   object
 5   discount_percentage  1465 non-null   object
 6   rating               1465 non-null   object
 7   rating_count         1463 non-null   object
 8   about_product        1465 non-null   object
 9   user_id              1465 non-null   object
 10  user_name            1465 non-null   object
 11  review_id            1465 non-null   object
 12  review_title         1465 non-null   object
 13  review_content       1465 non-null   object
 14  img_link             1465 non-null   object
 1

- We have 1465 rows(entries) and 16 columns(attributes).
- All of our attributes are of **object** Datatype, which means that even numerical values like discounted_price, actual_price, rating count, discounted_percentage, and rating are stored as text.

### **Data Preprocessing**

In [167]:
# Step 2: Data Preprocessing

# Data Cleaning

print('\nMissing Values:')
print(amazon_sales.isna().sum())


Missing Values:
product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           2
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64


- As we can see, we have **2** missing values in **rating_count** column out of **1465** so we are going to drop them because it makes nearly **0.14%** of total data, making it insignificant.

In [168]:
amazon_sales = amazon_sales.dropna()
print(amazon_sales.isna().sum())

product_id             0
product_name           0
category               0
discounted_price       0
actual_price           0
discount_percentage    0
rating                 0
rating_count           0
about_product          0
user_id                0
user_name              0
review_id              0
review_title           0
review_content         0
img_link               0
product_link           0
dtype: int64


In [169]:
# Check for duplicates
print("\nDuplicate Values:")
print(amazon_sales.duplicated().sum())


Duplicate Values:
0


- There are no duplicate entries in our dataset.

In [170]:
print(amazon_sales['rating'].unique())

# Remove | in 'rating' column
amazon_sales = amazon_sales[amazon_sales['rating'] != '|']

['4.2' '4.0' '3.9' '4.1' '4.3' '4.4' '4.5' '3.7' '3.3' '3.6' '3.4' '3.8'
 '3.5' '4.6' '3.2' '5.0' '4.7' '3.0' '2.8' '4' '3.1' '4.8' '2.3' '|' '2'
 '3' '2.6' '2.9']


In [171]:
# Convert columns to appropriate data types

amazon_sales['discounted_price'] = amazon_sales['discounted_price'].astype(str).str.replace('₹', '').str.replace(',', '').astype(float)
amazon_sales['actual_price'] = amazon_sales['actual_price'].astype(str).str.replace('₹', '').str.replace(',', '').astype(float)
amazon_sales['discount_percentage'] = amazon_sales['discount_percentage'].astype(str).str.replace('%', '').astype(float)
amazon_sales['rating'] = amazon_sales['rating'].astype(float)
amazon_sales['rating_count'] = amazon_sales['rating_count'].astype(str).str.replace(',','').astype(int)
amazon_sales.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1462 entries, 0 to 1464
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   product_id           1462 non-null   object 
 1   product_name         1462 non-null   object 
 2   category             1462 non-null   object 
 3   discounted_price     1462 non-null   float64
 4   actual_price         1462 non-null   float64
 5   discount_percentage  1462 non-null   float64
 6   rating               1462 non-null   float64
 7   rating_count         1462 non-null   int64  
 8   about_product        1462 non-null   object 
 9   user_id              1462 non-null   object 
 10  user_name            1462 non-null   object 
 11  review_id            1462 non-null   object 
 12  review_title         1462 non-null   object 
 13  review_content       1462 non-null   object 
 14  img_link             1462 non-null   object 
 15  product_link         1462 non-null   object

- We converted  **discounted_price, actual_price, discount_percentage,and rating** to float and **rating_count** to int from object so that numerical columns will not be considered as text.

In [172]:
# Basic stats
print("Summary Stats")
print(amazon_sales.describe())

Summary Stats
       discounted_price   actual_price  discount_percentage       rating  \
count       1462.000000    1462.000000          1462.000000  1462.000000   
mean        3129.981826    5453.087743            47.672367     4.096717   
std         6950.548042   10884.467444            21.613905     0.289497   
min           39.000000      39.000000             0.000000     2.000000   
25%          325.000000     800.000000            32.000000     4.000000   
50%          799.000000    1670.000000            50.000000     4.100000   
75%         1999.000000    4321.250000            63.000000     4.300000   
max        77990.000000  139900.000000            94.000000     5.000000   

        rating_count  
count    1462.000000  
mean    18307.376881  
std     42766.096572  
min         2.000000  
25%      1191.500000  
50%      5179.000000  
75%     17342.250000  
max    426973.000000  


 - Products range from **₹39 to ₹1,39,900**, with an average discount of **47.67%**, making them more affordable.
 - High average rating of **4.1/5** shows strong customer satisfaction.
 - Rating counts vary widely **(2 to 426,973)**, indicating some products are far more popular than others.
 - Discounts up to **94%** highlight aggressive pricing strategies to boost sales.

### **Exploratory Data Analysis (EDA)**

In [173]:
print('EDA with Plotly Visualizations.')

EDA with Plotly Visualizations.


Plotly is a Python library that creates interactive, web-based visualizations. It provides tools like Plotly Express for quick plots and Graph Objects for detailed customization.

#### **Q1:  What is the average rating for top 20 product categories?**

In [174]:
# Avg. rating for each product category
category_avg_rating = amazon_sales.groupby('category')['rating'].mean().reset_index() 

print("\nAverage rating for top 20 product categories")

# Avg. rating for top 20 Product categories
top_categories = category_avg_rating.nlargest(20, 'rating')  # Top 20 categories

top_categories


Average rating for top 20 product categories


category    rating
57                       Computers&Accessories|Tablets  4.600000
38             Computers&Accessories|Components|Memory  4.500000
48   Computers&Accessories|NetworkingDevices|Networ...  4.500000
62    Electronics|Cameras&Photography|Accessories|Film  4.500000
81   Electronics|HomeAudio|MediaStreamingDevices|St...  4.500000
118       Electronics|PowerAccessories|SurgeProtectors  4.500000
123      Home&Kitchen|CraftMaterials|PaintingMaterials  4.500000
148  Home&Kitchen|Kitchen&HomeAppliances|Coffee,Tea...  4.500000
175  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...  4.500000
194          HomeImprovement|Electrical|CordManagement  4.500000
196  OfficeProducts|OfficeElectronics|Calculators|B...  4.500000
107  Electronics|Mobiles&Accessories|MobileAccessor...  4.471429
155  Home&Kitchen|Kitchen&HomeAppliances|SmallKitch...  4.460000
198  OfficeProducts|OfficeElectronics|Calculators|S...  4.450000
124  Home&Kitchen|CraftMaterials|PaintingMaterials|...  4.433333
71   Electronics|GeneralPurposeBatteries&BatteryCha...  4.414286
7    Computers&Accessories|Accessories&Peripherals|...  4.400000
8    Computers&Accessories|Accessories&Peripherals|...  4.400000
41   Computers&Accessories|ExternalDevices&DataStor...  4.400000
59   Electronics|Accessories|MemoryCards|SecureDigi...  4.400000

In [175]:
# bar plot for average rating for each product category
fig1 = px.bar(data_frame=top_categories, 
              x='rating', 
              y='category', 
              title='Average Rating By Top 20 Product Categories',
              orientation='h',
              color='category')  # Horizontal bar chart

fig1.update_layout(xaxis_title='Average Rating', yaxis_title='Product Category', height=800, showlegend=False)
fig1.show()

- The **average rating** for the **top 20 product categories** ranges from **4.4 to 4.6**, showing high customer satisfaction across these categories.
- Tech products, home essentials, and office supplies dominating, reflecting strong customer satisfaction and market demand. 

#### **Q2: What are the top rated products by category?**

In [176]:
print("\nTop rated products by category")

top_products_by_category = amazon_sales.groupby(['category', 'product_name'])['rating_count'].max().reset_index()
top_products_by_category = top_products_by_category.sort_values(by='rating_count', ascending=False).head(10)
top_products_by_category


Top rated products by category


category  \
500  Electronics|HomeTheater,TV&Video|Accessories|C...   
505  Electronics|HomeTheater,TV&Video|Accessories|C...   
499  Electronics|HomeTheater,TV&Video|Accessories|C...   
459  Electronics|Headphones,Earbuds&Accessories|Hea...   
458  Electronics|Headphones,Earbuds&Accessories|Hea...   
456  Electronics|Headphones,Earbuds&Accessories|Hea...   
749  Electronics|Mobiles&Accessories|Smartphones&Ba...   
747  Electronics|Mobiles&Accessories|Smartphones&Ba...   
748  Electronics|Mobiles&Accessories|Smartphones&Ba...   
750  Electronics|Mobiles&Accessories|Smartphones&Ba...   

                                          product_name  rating_count  
500  Amazon Basics High-Speed HDMI Cable, 6 Feet - ...        426973  
505  AmazonBasics Flexible Premium HDMI Cable (Blac...        426973  
499  Amazon Basics High-Speed HDMI Cable, 6 Feet (2...        426973  
459  boAt Bassheads 100 in Ear Wired Earphones with...        363713  
458  boAt Bassheads 100 in Ear Wired Earphones with...        363713  
456  boAt BassHeads 100 in-Ear Wired Headphones wit...        363711  
749  Redmi 9A Sport (Coral Green, 2GB RAM, 32GB Sto...        313836  
747  Redmi 9 Activ (Carbon Black, 4GB RAM, 64GB Sto...        313836  
748  Redmi 9A Sport (Carbon Black, 2GB RAM, 32GB St...        313832  
750  Redmi 9A Sport (Coral Green, 3GB RAM, 32GB Sto...        313832

In [177]:
# bar plot for top rated product by category

fig2 = px.bar(data_frame=top_products_by_category,
              x='rating_count',
              y='product_name',
              color='category',
              title='Top Rated Products by Category',
              orientation='h',
              category_orders={"product_name": top_products_by_category["product_name"].tolist()}
              )

fig2.update_layout(yaxis_title='Product Name', xaxis_title='Rating Count',height= 600, showlegend=True, legend=dict(
        title="Categories",
        orientation="h",  # Horizontal legend
        yanchor="top",  # Align to top
        y=-0.2,  # Move legend below the plot
        xanchor="center",  # Center align
        x=-0.2 
    )
)
fig2.show()

- **Amazon Basics HDMI cables(426.973k)**, **boAt Bassheads earphones(363.713k)**, and **Redmi 9A Sport/9 Activ smartphones(313.836k)** are the top-rated products in their respective categories.
- High Rating count suggests strong demand for these products.

#### **Q3: What is the distribution of discounted prices vs. actual prices?**

In [184]:
print('\nDistribution of discounted prices vs. actual prices')

fig3 = px.histogram(amazon_sales, x=['discounted_price', 'actual_price'], barmode='overlay', 
                    title='Distribution of Discounted Prices vs. Actual Prices')
fig3.update_layout(xaxis_title='Price', yaxis_title='Frequency', legend=dict(title="Prices"))
fig3.show()



Distribution of discounted prices vs. actual prices


- The histogram shows that most products have **lower prices**, with both **actual** and **discounted** prices skewed towards the **lower end**. 
- Discounted prices (blue) are generally **lower** than actual prices (red), indicating that products are generally being sold at a discount.
- Few high-priced products exist, but they are rare. 
- The overlap suggests some items have minimal or no discounts.

#### **Q4: How does the average discount percentage vary across the top 10 discounted categories??**

In [193]:
category_avg_discount = amazon_sales.groupby('category')['discount_percentage'].mean().reset_index()
top_discounted_categories = category_avg_discount.nlargest(10, 'discount_percentage').sort_values(by='discount_percentage', ascending=False)
top_discounted_categories

category  discount_percentage
6    Computers&Accessories|Accessories&Peripherals|...                 90.0
75   Electronics|Headphones,Earbuds&Accessories|Ear...                 90.0
106  Electronics|Mobiles&Accessories|MobileAccessor...                 90.0
73   Electronics|Headphones,Earbuds&Accessories|Ada...                 88.0
14   Computers&Accessories|Accessories&Peripherals|...                 87.5
110  Electronics|Mobiles&Accessories|MobileAccessor...                 82.0
36   Computers&Accessories|Components|InternalHardD...                 80.0
1    Computers&Accessories|Accessories&Peripherals|...                 78.5
114  Electronics|Mobiles&Accessories|MobileAccessor...                 76.4
74    Electronics|Headphones,Earbuds&Accessories|Cases                 76.0

In [ ]:
# bar chart for how average discount percentage varies across top 10 discounted categories
fig4 = px.bar(top_discounted_categories, x='discount_percentage', y='category', 
              title='Top 10 Categories with Highest Average Discounts', orientation='h', color='category')
fig4.update_layout(xaxis_title='Average Discount Percentage (%)', yaxis_title='Category',height=800, showlegend=False)
fig4.show()

- **Top 3 discounted categories (90%):** Cable Connection Protectors, Earpads, Phone Charms.
- **Electronics & Accessories** dominate, with all top 10 categories having **75%+ discounts**.
- Likely clearance sales or bundling strategy to boost purchases.